### PR TITLE
Implement the containerd runtime for Ignite

### PR DESCRIPTION
**Pull request by twelho**
This PR implements the alternative `containerd` runtime for Ignite. containerd `v1.2.7`, the latest stable version, was automatically vendored, but this bumps that to `v1.3.0-beta.1` for multiple reasons:
- Image exporting is not possible with `v1.2.7`, the old extractor was already dropped in that release before transitioning to the new implementation for `v1.3.0`
- `attach` is natively implemented, this also enables `logs`
- There is a new fully featured CLI for containerd I can use for implementation reference

The runtime interface has also received some changes to better adapt to both container runtimes.

Fixes https://github.com/weaveworks/ignite/issues/209.

cc @luxas

### CODE REVIEW

**Comment by luxas** [comment]
Also see https://github.com/containerd/containerd/blob/master/docs/getting-started.md for a guide

**Comment by twelho** [comment]
The containerd vendoring has been separated in https://github.com/weaveworks/ignite/pull/348, this PR has now been rebased on top of it.

File: pkg/runtime/containerd/client.go [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+
+func (cc *ctdClient) PullImage(image string) (io.ReadCloser, error) {
+	_, err := cc.client.Pull(context.Background(), image) // TODO: RemoteOptions
+	return nil, err
+}
+
+func (cc *ctdClient) ExportImage(image string) (io.ReadCloser, string, error) {
+	r, w := io.Pipe()
+	err := cc.client.Export(context.Background(), w)
+	return r, "", err // We don't need a temporary container, leave the ID string blank
```
[the remark this answers is not in the document]
Comment (twelho) [reply pkg/runtime/containerd/client.go]: It does already support a blank string there. I might refactor the interface to handle a specific `ContainerID` type which can be `nil` so we can avoid confusion.

File: pkg/runtime/containerd/client.go [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+		ID:        info.ID,
+		Image:     info.Image, // TODO: This may be incorrect
+		Status:    "",         // TODO: This
+		IPAddress: nil,        // TODO: This
+	}, nil
+}
+
+func (cc *ctdClient) AttachContainer(container string) error {
+	// TODO: Implement attach for containerd
+	return unsupported("logs")
```
[the remark this answers is not in the document]
Comment (twelho) [reply pkg/runtime/containerd/client.go]: Fixed, thanks

File: pkg/runtime/containerd/client.go [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+
+	"github.com/containerd/containerd"
+	"github.com/containerd/containerd/cio"
+	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
+	"github.com/weaveworks/ignite/pkg/constants"
+	"github.com/weaveworks/ignite/pkg/runtime"
+)
+
+const (
+	ctdSocket    = constants.DATA_DIR + "/containerd.sock"
```
[the remark this answers is not in the document]
Comment (twelho) [reply pkg/runtime/containerd/client.go]: Changed, it seems that there's no way to ask `containerd` where the socket is, so we might want to make this configurable in the future.

File: pkg/runtime/types.go [hunk 4]
```diff
@@ -49,7 +49,7 @@ type ContainerConfig struct {
 
 type Interface interface {
 	InspectImage(image string) (*ImageInspectResult, error)
-	PullImage(image string) (io.ReadCloser, error)
+	PullImage(image string) (io.ReadCloser, error) // TODO: This doesn't need to return the io.ReadCloser
```
[the remark this answers is not in the document]
Comment (twelho) [reply pkg/runtime/types.go]: :+1:

File: pkg/runtime/containerd/client.go [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+	}, nil
+}
+
+func (cc *ctdClient) AttachContainer(container string) error {
+	// TODO: Implement attach for containerd
+	return unsupported("logs")
+}
+
+func (cc *ctdClient) RunContainer(image string, config *runtime.ContainerConfig, name string) (s string, err error) {
+	img, err := cc.client.GetImage(context.Background(), image)
```
[the remark this answers is not in the document]
Comment (twelho) [reply pkg/runtime/containerd/client.go]: Solved using `containerd.WithPullUnpack` during pulls

**Comment by twelho** [comment]
Rebased on top of https://github.com/weaveworks/ignite/pull/356.

**Comment by luxas** [comment]
I implemented RunContainer now based off previous PoC code I had created

**Comment by twelho** [comment]
The `containerd` runtime now has basic features complete and supports a full `ignite run` from scratch. It still needs work to be able to stop/remove VMs and to support the rest of the `runtime` features, but those improvements shall be implemented in separate PRs, this one is already quite massive. The `containerd` runtime is not enabled by default (as it's still incomplete).

**Comment by luxas** [comment]
Awesome work :tada: :sparkles:!!
Let's get this going

**Pull request merged**